pages/_app.tsx

```diff
@@ -51,10 +54,51 @@ if (process.browser) {
   Storage.init();
 }
 
-export default class App extends NextApp {
+interface AppState {
+  precached: boolean;
+  error?: Error;
+}
+
+export default class App extends NextApp<{}, {}, AppState> {
+  state: AppState = { precached: false };
+
+  private isUnmounted = true;
+
+  async componentDidMount(): Promise<void> {
+    this.isUnmounted = false;
+
+    const stateUpdate: Partial<AppState> = {};
+
+    try {
+      const precached = await this.precacheAll();
+
+      stateUpdate.precached = precached;
+    } catch (err) {
+      stateUpdate.error = err;
+    }
+
+    if (this.isUnmounted) {
+      return;
+    }
+
+    this.setState(state => ({ ...state, ...stateUpdate }));
+  }
+
+  shouldComponentUpdate(nextProps: {}, nextState: AppState): boolean {
+    return !nextState.precached === true;
+  }
```

Comment:
There is probably a smarter way of writing `false if the only thing in nextState is {precached: true}`

Reply:
I think this is generally wrong. What if precached changed at the same time as something else? I think it's worth installing https://github.com/FormidableLabs/react-fast-compare. Then you end up with something like:

```ts
return !isEqual(this.props, nextProps) || !isEqual({ ...this.state, precached: undefined }, { ...nextState, precached: undefined });
```